### PR TITLE
Make the websocket auto reconnect

### PR DESCRIPTION
**Pull request by uellenberg**
I haven't tested this on every kind of failure, but it should be solid enough to handle them. My main concern is that, if latency is too large, we'll get stuck in a loop of continuously reconnecting. This should be solvable by just increasing the disconnect time, maybe making it configurable somewhere on the web page just in case.

### CODE REVIEW

**Comment by adamseth2** [comment]
Not sure why but I get the continuous reconnection loops even though I increase disconnect time (Below is what the urc repo keeps repeating): 
```
[rosbridge_websocket-1] [INFO] [1729506972.850095159] [rosbridge_websocket]: Client disconnected. 0 clients total.
[rosbridge_websocket-1] [INFO] [1729506973.516280020] [rosbridge_websocket]: Calling services in existing thread
[rosbridge_websocket-1] [INFO] [1729506973.518389495] [rosbridge_websocket]: Client connected. 1 clients total.
```

**Comment by uellenberg** [comment]
> Not sure why but I get the continuous reconnection loops even though I increase disconnect time (Below is what the urc repo keeps repeating):
> 
> ```
> [rosbridge_websocket-1] [INFO] [1729506972.850095159] [rosbridge_websocket]: Client disconnected. 0 clients total.
> [rosbridge_websocket-1] [INFO] [1729506973.516280020] [rosbridge_websocket]: Calling services in existing thread
> [rosbridge_websocket-1] [INFO] [1729506973.518389495] [rosbridge_websocket]: Client connected. 1 clients total.
> ```

We've confirmed this is due to older versions of rosbridge_server not supporting COBR. If anyone comes across this issue, you need to update your rosbridge_server version (ideally by rebuilding the container WITHOUT CACHE).

**Comment by adamseth2** [comment]
Debugged problem. Once Kyler approves the PQ on the rover code, please merge.

File: src/roslib/rosInit.ts [hunk 1]
```diff
@@ -1,24 +1,82 @@
...
+      // Forcefully reconnect.
+      ros.socket = null;
+      // Give the reconnection extra time so that it doesn't
+      // immediately get killed.
+      heartbeatTime = Date.now() + reconnectionGraceSeconds * secondsToTimestamp;
+      ros.connect(serverHost);
+    }
+  }, 100);
+
   return { ros, isWebSocketConnected };
```
Comment (adamseth2) [inline src/roslib/rosInit.ts:81]: Did not return stop, Causing Typescirpt error

File: src/roslib/rosInit.ts [hunk 2]
```diff
@@ -1,24 +1,82 @@
...
+    if (Date.now() - heartbeatTime > heartbeatDisconnectSeconds * secondsToTimestamp) {
+      isWebSocketConnected.value = false;
+
+      // No need to error if the websocket is already closed.
+      try {
+        ros.close();
+      } catch (_) {}
+
+      // Forcefully reconnect.
+      ros.socket = null;
```
Comment (adamseth2) [inline src/roslib/rosInit.ts:73]: Gives this Typescript Error:

If necessary, edit the roslib@types file and patch it. 

`Property 'socket' does not exist on type 'Ros'.ts(2339)`

**Review by adamseth2 — APPROVED** [review]
Also after addressed issues, feel free to merge

**Pull request merged**